### PR TITLE
ceph: raw mode OSD on LV-backed PVC

### LINKED ISSUE #5939: ceph: support raw mode OSD on LV-backed PVC
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

Support raw mode OSD on LV-backed PVC. Although OSD on PVC uses raw mode, LV-backed PVC still use LVM mode.

**What is use case behind this feature:**

It can simplify OSD management in many ways and can avoid many problems that come from LVM. For example, in OSD on LV-backed PVC, which uses LVM mode OSD, LV's tag may be updated from both host and container simultaneously. The former is from admin in the host, the latter is from Rook/Ceph container to add Ceph-related tags. As a result, LVM metadata may corrupt.

Added(Aug 31, 2020):
Fore more detail, please refer to the following document.
https://github.com/rook/rook/blob/master/Documentation/ceph-common-issues.md#lvm-metadata-can-be-corrupted-with-osd-on-lv-backed-pvc

### PR DESCRIPTION
**Pull request by llamerada-jp**
**Description of your changes:**

Use raw mode OSD instead of lvm mode even if LV-backed PVC.

**Which issue is resolved by this Pull Request:**

Resolves #5939 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test full]

### CODE REVIEW

**Comment by llamerada-jp** [comment]
@travisn 
We're going to announce users how to resolve #5939 as follows. How about this?

- Describe how to resolve this issue in [`Ceph Common Issues`](https://rook.io/docs/rook/v1.4/ceph-common-issues.html#lvm-metadata-can-be-corrupted-with-osd-on-lv-backed-pvc).
- Point this description from [`Considerations` in `ceph-upgrade.md`](https://github.com/rook/rook/blob/master/Documentation/ceph-upgrade.md), and [`PendingReleaseNotes.md`](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md#upgrade-guides).
- Notify it in slack.

**Comment by travisn** [comment]
@llamerada-jp The CI is being skipped because the PR is still marked as in progress. Is it about ready for review? If it's getting close, I'd recommend changing the PR status so the CI will run and give you the validation.

> We're going to announce users how to resolve #5939 as follows. How about this?
> 
> * Describe how to resolve this issue in [`Ceph Common Issues`](https://rook.io/docs/rook/v1.4/ceph-common-issues.html#lvm-metadata-can-be-corrupted-with-osd-on-lv-backed-pvc).

Are you adding more to that topic? Or it already has the needed comments?

> * Point this description from [`Considerations` in `ceph-upgrade.md`](https://github.com/rook/rook/blob/master/Documentation/ceph-upgrade.md), and [`PendingReleaseNotes.md`](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md#upgrade-guides).

@llamerada-jp In the upgrade guide are you going to recommend that users convert to the raw OSDs? But as long as the existing OSDs still run ok, this could be an optional step to take so their existing OSDs wouldn't be blocked. 

> * Notify it in slack.

This type of change would be best if it is with the minor release. During the patch releases we try to avoid breaking changes. In slack, I assume we could just mention the recommendation that is coming in 1.5. If we backport this change to 1.4, it will help them start transitioning sooner if they want.

**Comment by llamerada-jp** [comment]
@travisn

The changes to the body of the program have been made, so I changed status to ready for review.

> Are you adding more to that topic? Or it already has the needed comments?
Yes, I'm going to add a procedure for convert to the raw OSD.

> In the upgrade guide are you going to recommend that users convert to the raw OSDs?
Yes, it's not "must" but we recommend convert to the raw OSDs because this issue can corrupt existing lv OSDs.

> I assume we could just mention the recommendation that is coming in 1.5. If we backport this change to 1.4, it will help them start transitioning sooner if they want.
Thank you, I will.

**Comment by llamerada-jp** [comment]
@travisn 

We added the documentations we thought we needed. 
Completely ready for review now 😃

File: Documentation/ceph-common-issues.md
Comment (travisn) [inline Documentation/ceph-common-issues.md:821]: Existing lvm mode OSDs will still be supported during upgrade, right? The support being removed looks like it is for creating new OSDs. That support is only being removed from the osd prepare job. So it appears the existing OSDs will still be upgraded and are expected to be fully functional. Let's clarify that existing OSDs will continue to work after the upgrade.

**Review by travisn — CHANGES_REQUESTED** [review]

File: PendingReleaseNotes.md
Comment (llamerada-jp) [reply PendingReleaseNotes.md]: I reflected your comment.

File: Documentation/ceph-common-issues.md
Comment (llamerada-jp) [reply Documentation/ceph-common-issues.md]: I reflected your comment.

File: pkg/daemon/ceph/osd/volume.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume.go:113]: We think so.

File: pkg/daemon/ceph/osd/volume.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume.go:125]: Yes, it is.

**Review by travisn — CHANGES_REQUESTED** [review]
@llamerada-jp  Just to be very clear, Rook will always create all new OSDs in raw mode now, correct? 
@leseb Were we waiting for a fix in ceph-volume to support this? I thought we were blocked on doing this. But the integration tests are passing, so it seems to be working on the latest nautilus and octopus.

File: Documentation/ceph-common-issues.md
Comment (satoru-takeuchi) [reply Documentation/ceph-common-issues.md:821]: > Just to be very clear, Rook will always create all new OSDs in raw mode now, correct?

@travisn Yes in OSD on PVC.

> Were we waiting for a fix in ceph-volume to support this? I thought we were blocked on doing this. But the integration tests are passing, so it seems to be working on the latest nautilus and octopus.

Currently, `ceph-volume raw prepare` can create LVM mode OSD. However, `ceph-volume inventory` regards this LVM as unavailable. I'm working on fixing this problem. As a first step, the following PR will support crypt.

https://github.com/ceph/ceph/pull/34966

After merging this PR, I'll try to support LVM in raw mode OSD. Before that, Rook will call `ceph-volume raw prepare` without calling `ceph-volume inventory` as current cyrpt and mpath.

Comment (travisn) [reply Documentation/ceph-common-issues.md:821]: Ok, I was confusing removing the "ceph-volume lvm" mode with detecting whether the PV is backed by LV. This PR is only changing OSD creation when on PVs that are backed by an LV so that we always use raw mode in that case, right? So this isn't as big a switch for all scenarios as I was thinking.

Comment (satoru-takeuchi) [reply Documentation/ceph-common-issues.md:821]: > This PR is only changing OSD creation when on PVs that are backed by an LV so that we always use raw mode in that case, right? So this isn't as big a switch for all scenarios as I was thinking.

Correct.

File: pkg/daemon/ceph/osd/volume.go
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go:125]: @llamerada-jp ?

Comment (satoru-takeuchi) [reply pkg/daemon/ceph/osd/volume.go:125]: @leseb `lvPath `is passed iff `pvcBacked` is `true`, and then there is only one OSD per PV(LogicalVolume). So `GetCephVolumeLVMOSDs` return at most one OSD in this case.

Correct me if I misunderstood what you said.

Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go:125]: I guess I don't understand why we need to escape if `en(osds) > 0`, shouldn't it be at least `> 1`?

File: Documentation/ceph-common-issues.md
Comment (travisn) [reply Documentation/ceph-common-issues.md:822]: This statement still doesn't seem correct:
"Rook no longer uses `ceph-volume` lvm mode in a prepare job and creates OSDs with `ceph-volume` raw mode instead after merging the [PR](https://github.com/rook/rook/pull/6184)."

There are still some scenarios where lvm mode is used, such as for encrypted clusters.  I'd also suggest not mentioning the PR number, it doesn't matter after the merge.

File: pkg/daemon/ceph/osd/volume.go
Comment (travisn) [reply pkg/daemon/ceph/osd/volume.go:125]: @leseb If it is `> 0` the osd already exists and won't be created again. If it `= 0`, the osd will be created.

File: Documentation/ceph-common-issues.md
Comment (satoru-takeuchi) [reply Documentation/ceph-common-issues.md]: Thanks, fixed.

**Comment by satoru-takeuchi** [comment]
@travisn Although the review of this PR is almost finished, there seem to be many intermittent test failures that probably aren't related to this PR. So we'll try to merge it after fixing these. Please let me know if you have any other ideas.

**Comment by travisn** [comment]
> @travisn Although the review of this PR is almost finished, there seem to be many intermittent test failures that probably aren't related to this PR. So we'll try to merge it after fixing these. Please let me know if you have any other ideas.

Since [test full] is added, we will run all of the ceph test suites on every version of K8s, just like in master and release builds. With this enabled, it is currently unlikely to get a fully green build. In the last build, there were 4 out of 5 green K8s versions, which is a very positive indication that there are no regressions in the tests. The one failed test does look like a known issue. For now, I restarted the CI to get another test run for you.

**Comment by satoru-takeuchi** [comment]
Unfortunately, CI failed again. I created the following issue that tracks the intermittent failures. I'll try to fix #6358 and re-run CI once in a while.

https://github.com/rook/rook/issues/6358

**Comment by satoru-takeuchi** [comment]
memo:
CI of aws_1.16.x is failed with pattern `C` of #6358`. This problem would be fixed by #6388 . Other tests passed.

**Comment by satoru-takeuchi** [comment]
@leseb @travisn Finally this PR passed the CI and we reflected all comments. Could you take a look at this again?

File: pkg/daemon/ceph/osd/volume_test.go
Comment (leseb) [inline pkg/daemon/ceph/osd/volume_test.go:643]: Isn't this the same as the previous test? 👀

File: pkg/daemon/ceph/osd/volume_test.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume_test.go]: I fixed the comment.

File: pkg/daemon/ceph/osd/volume_test.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume_test.go:422]: I fixed comment. There are a OSD prepared with RAW mode.

File: pkg/daemon/ceph/osd/volume_test.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume_test.go:473]: Fixed comment too. There is a  OSD prepared with LVM mode.

File: pkg/daemon/ceph/osd/volume_test.go
Comment (llamerada-jp) [reply pkg/daemon/ceph/osd/volume_test.go]: Sorry. It is the same test and I delete a duplicate test.

**Review by leseb — APPROVED** [review]
LTGM!

**Comment by satoru-takeuchi** [comment]
@leseb CI passed. Please merge it.

**Pull request merged**
